customize toString() on enums

Hi,

With the ToString plugin, classes implements ToString but not enums.
With the EnumValue plugin, enums can EnumValue.

Maybe I didn't find how to do, but is it possible to customize the toString method of enums? For exemple to use a resourceBundle.

Best regards,

I tested the PR and it works like a charm!

Just notice all tests didn't pass on my computer. But that's something I have on master also.
Do you want I add a new issue about this? I checked on Windows 10 with JDK 21.

> [INFO] Running org.jvnet.jaxb.maven.util.tests.CollectionsUtilsTests
> [INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 s -- in org.jvnet.jaxb.maven.util.tests.CollectionsUtilsTests
> [INFO]
> [INFO] Results:
> [INFO]
> [ERROR] Failures:
> [ERROR]   IOUtilsTests.reorderFilesIncludesWithWildcardFirst:76 expected:<[common/c1].xsd> but was:<[a].xsd>
> [ERROR]   IOUtilsTests.reorderFilesIncludesWithWildcardMiddle:99 expected:<[common/c1].xsd> but was:<[b].xsd>
> [INFO]
> [ERROR] Tests run: 14, Failures: 2, Errors: 0, Skipped: 0

--- a/jaxb-plugins-parent/jaxb-plugins-runtime/src/main/java/org/jvnet/jaxb/lang/EnumToStringStrategy.java
+++ b/jaxb-plugins-parent/jaxb-plugins-runtime/src/main/java/org/jvnet/jaxb/lang/EnumToStringStrategy.java
@@ -1,0 +1,31 @@
+package org.jvnet.jaxb.lang;
+
+import org.jvnet.jaxb.locator.ObjectLocator;
+
+/**
+ * Original submission by Kevin Senechal (kevin.senechal@dooapp.com) on 30/07/15.
+ */
+public class EnumToStringStrategy extends JAXBToStringStrategy {
+
+    public StringBuilder appendStart(ObjectLocator parentLocator, Object parent, StringBuilder stringBuilder) {
+        if (parent instanceof Enum) {
+            return stringBuilder;
+        } else {
+            return super.appendStart(parentLocator, parent, stringBuilder);
+        }
+    }
+
+    public StringBuilder appendEnd(ObjectLocator parentLocator, Object parent, StringBuilder stringBuilder) {
+        if (parent instanceof Enum) {
+            return appendEnum((Enum) parent, stringBuilder);
+        } else {
+            return super.appendEnd(parentLocator, parent, stringBuilder);
+        }
+    }
+
+    protected StringBuilder appendEnum(Enum e, StringBuilder stringBuilder) {
+        stringBuilder.append(e.name());
+        return stringBuilder;
+    }
+
+}

--- a/jaxb-plugins-parent/jaxb-plugins/src/main/java/org/jvnet/jaxb/plugin/tostring/ToStringPlugin.java
+++ b/jaxb-plugins-parent/jaxb-plugins/src/main/java/org/jvnet/jaxb/plugin/tostring/ToStringPlugin.java
@@ -2,9 +2,12 @@ package org.jvnet.jaxb.plugin.tostring;
 
 import java.util.Arrays;
 import java.util.Collection;
+import java.util.Map;
 
 import javax.xml.namespace.QName;
 
+import com.sun.codemodel.JFieldVar;
+import com.sun.tools.xjc.outline.EnumOutline;
 import org.jvnet.jaxb.lang.JAXBToStringStrategy;
 import org.jvnet.jaxb.lang.ToString;
 import org.jvnet.jaxb.lang.ToStringStrategy;
@@ -57,7 +60,7 @@ public class ToStringPlugin extends AbstractParameterizablePlugin {
 		this.fieldAccessorFactory = fieldAccessorFactory;
 	}
 
-	private String toStringStrategyClass = JAXBToStringStrategy.class.getName();
+	private String toStringStrategyClass = JAXBToStringStrategy.class.getName();;
 
 	public void setToStringStrategyClass(String toStringStrategy) {
 		this.toStringStrategyClass = toStringStrategy;
@@ -67,7 +70,17 @@ public class ToStringPlugin extends AbstractParameterizablePlugin {
 		return toStringStrategyClass;
 	}
 
-	public JExpression createToStringStrategy(JCodeModel codeModel) {
+    private boolean toStringEnums = false;
+
+    public void setToStringEnums(boolean toStringEnums) {
+        this.toStringEnums = toStringEnums;
+    }
+
+    public boolean isToStringEnums() {
+        return toStringEnums;
+    }
+
+    public JExpression createToStringStrategy(JCodeModel codeModel) {
 		return StrategyClassUtils.createStrategyInstanceExpression(codeModel,
 				ToStringStrategy.class, getToStringStrategyClass());
 	}
@@ -102,32 +115,48 @@ public class ToStringPlugin extends AbstractParameterizablePlugin {
 			            org.jvnet.jaxb.plugin.LegacyCustomizations.GENERATED_ELEMENT_NAME);
 	}
 
-	@Override
-	public boolean run(Outline outline, Options opt, ErrorHandler errorHandler) {
-		for (final ClassOutline classOutline : outline.getClasses())
-			if (!getIgnoring().isIgnored(classOutline)) {
-				processClassOutline(classOutline);
-			}
-		return true;
-	}
+    @Override
+    public boolean run(Outline outline, Options opt, ErrorHandler errorHandler) {
+        for (final ClassOutline classOutline : outline.getClasses()) {
+            if (!getIgnoring().isIgnored(classOutline)) {
+                processClassOutline(classOutline);
+            }
+        }
+        if (isToStringEnums()) {
+            for (final EnumOutline enumOutline : outline.getEnums()) {
+                if (!getIgnoring().isIgnored(enumOutline)) {
+                    processEnumOutline(enumOutline);
+                }
+            }
+        }
+        return true;
+    }
 
 	protected void processClassOutline(ClassOutline classOutline) {
 		final JDefinedClass theClass = classOutline.implClass;
 		ClassUtils._implements(theClass, theClass.owner().ref(ToString.class));
 
 		@SuppressWarnings("unused")
-		final JMethod object$toString = generateObject$toString(classOutline,
-				theClass);
+		final JMethod object$toString = generateObject$toString(theClass);
 		@SuppressWarnings("unused")
-		final JMethod toString$append = generateToString$append(classOutline,
-				theClass);
+		final JMethod toString$append = generateToString$append(theClass);
 		@SuppressWarnings("unused")
-		final JMethod toString$appendFields = generateToString$appendFields(
-				classOutline, theClass);
+		final JMethod toString$appendFields = generateToString$appendFields(classOutline, theClass);
 	}
 
-	protected JMethod generateObject$toString(final ClassOutline classOutline,
-			final JDefinedClass theClass) {
+    protected void processEnumOutline(EnumOutline enumOutline) {
+        final JDefinedClass theClass = enumOutline.clazz;
+        ClassUtils._implements(theClass, theClass.owner().ref(ToString.class));
+
+        @SuppressWarnings("unused")
+        final JMethod object$toString = generateObject$toString(theClass);
+        @SuppressWarnings("unused")
+        final JMethod toString$append = generateToString$append(theClass);
+        @SuppressWarnings("unused")
+        final JMethod toString$appendFields = generateToString$appendFields(enumOutline, theClass);
+    }
+
+	protected JMethod generateObject$toString(final JDefinedClass theClass) {
 		final JCodeModel codeModel = theClass.owner();
 		final JMethod object$toString = theClass.method(JMod.PUBLIC,
 				codeModel.ref(String.class), "toString");
@@ -150,8 +179,7 @@ public class ToStringPlugin extends AbstractParameterizablePlugin {
 		return object$toString;
 	}
 
-	protected JMethod generateToString$append(final ClassOutline classOutline,
-			final JDefinedClass theClass) {
+	protected JMethod generateToString$append(final JDefinedClass theClass) {
 		final JCodeModel codeModel = theClass.owner();
 		final JMethod toString$append = theClass.method(JMod.PUBLIC,
 				codeModel.ref(StringBuilder.class), "append");
@@ -241,4 +269,23 @@ public class ToStringPlugin extends AbstractParameterizablePlugin {
 		return toString$appendFields;
 	}
 
+    protected JMethod generateToString$appendFields(EnumOutline enumOutline,
+                                                    final JDefinedClass theClass) {
+        final JCodeModel codeModel = theClass.owner();
+
+        final JMethod toString$appendFields = theClass.method(JMod.PUBLIC,
+            codeModel.ref(StringBuilder.class), "appendFields");
+        {
+            final JVar locator = toString$appendFields.param(
+                ObjectLocator.class, "locator");
+            final JVar buffer = toString$appendFields.param(
+                StringBuilder.class, "buffer");
+            final JVar toStringStrategy = toString$appendFields.param(
+                ToStringStrategy.class, "strategy");
+            final JBlock body = toString$appendFields.body();
+
+            body._return(buffer);
+        }
+        return toString$appendFields;
+    }
 }

--- a/jaxb-plugins-parent/tests/enumtostring/pom.xml
+++ b/jaxb-plugins-parent/tests/enumtostring/pom.xml
@@ -1,0 +1,47 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <groupId>org.jvnet.jaxb</groupId>
+    <artifactId>jaxb-plugins-tests</artifactId>
+    <version>4.0.6-SNAPSHOT</version>
+  </parent>
+  <artifactId>jaxb-plugins-test-enumtostring</artifactId>
+  <packaging>jar</packaging>
+  <name>JAXB Tools :: JAXB Plugins :: Test [enum-toString]</name>
+  <dependencies>
+    <dependency>
+      <groupId>org.jvnet.jaxb</groupId>
+      <artifactId>jaxb-maven-plugin-testing</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.jvnet.jaxb</groupId>
+      <artifactId>jaxb-plugins-runtime</artifactId>
+    </dependency>
+  </dependencies>
+  <build>
+    <defaultGoal>test</defaultGoal>
+    <plugins>
+      <plugin>
+        <groupId>org.jvnet.jaxb</groupId>
+        <artifactId>jaxb-maven-plugin</artifactId>
+        <configuration>
+          <extension>true</extension>
+          <args>
+            <arg>-XtoString</arg>
+            <arg>-XtoString-toStringStrategyClass=org.jvnet.jaxb.tests.enumtostring.MyEnumToStringStrategy</arg>
+            <arg>-XtoString-toStringEnums=true</arg>
+          </args>
+          <plugins>
+            <plugin>
+              <groupId>org.jvnet.jaxb</groupId>
+              <artifactId>jaxb-plugins</artifactId>
+            </plugin>
+          </plugins>
+        </configuration>
+      </plugin>
+    </plugins>
+  </build>
+</project>

--- a/jaxb-plugins-parent/tests/enumtostring/src/main/java/org/jvnet/jaxb/tests/enumtostring/MyEnumToStringStrategy.java
+++ b/jaxb-plugins-parent/tests/enumtostring/src/main/java/org/jvnet/jaxb/tests/enumtostring/MyEnumToStringStrategy.java
@@ -1,0 +1,15 @@
+package org.jvnet.jaxb.tests.enumtostring;
+
+import generated.Model;
+import org.jvnet.jaxb.lang.EnumToStringStrategy;
+
+public class MyEnumToStringStrategy extends EnumToStringStrategy {
+    @Override
+    protected StringBuilder appendEnum(Enum e, StringBuilder stringBuilder) {
+        if (e instanceof Model) {
+            stringBuilder.append(((Model) e).value());
+            return stringBuilder;
+        }
+        return super.appendEnum(e, stringBuilder);
+    }
+}

--- a/jaxb-plugins-parent/tests/enumtostring/src/main/resources/sample.xsd
+++ b/jaxb-plugins-parent/tests/enumtostring/src/main/resources/sample.xsd
@@ -1,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified">
+  <xs:element name="console">
+    <xs:complexType>
+      <xs:attribute name="model" type="model" use="required"/>
+      <xs:attribute name="vendor" type="vendor" use="required"/>
+    </xs:complexType>
+  </xs:element>
+
+  <xs:simpleType name="model">
+    <xs:restriction base="xs:string">
+      <xs:enumeration value="PS3"/>
+      <xs:enumeration value="PS4"/>
+      <xs:enumeration value="PS5"/>
+      <xs:enumeration value="XBOX360"/>
+      <xs:enumeration value="XBOXONE"/>
+    </xs:restriction>
+  </xs:simpleType>
+  <xs:simpleType name="vendor">
+    <xs:restriction base="xs:string">
+      <xs:enumeration value="SONY"/>
+      <xs:enumeration value="MICROSOFT"/>
+    </xs:restriction>
+  </xs:simpleType>
+</xs:schema>

--- a/jaxb-plugins-parent/tests/enumtostring/src/test/java/org/jvnet/jaxb/tests/enumtostring/ConsoleTest.java
+++ b/jaxb-plugins-parent/tests/enumtostring/src/test/java/org/jvnet/jaxb/tests/enumtostring/ConsoleTest.java
@@ -1,0 +1,33 @@
+package org.jvnet.jaxb.tests.enumtostring;
+
+import generated.Console;
+import generated.Model;
+import generated.Vendor;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class ConsoleTest {
+
+    @Test
+    public void testConsole() {
+        Console c = new Console();
+        c.setModel(Model.PS_5);
+        c.setVendor(Vendor.SONY);
+
+        String cToString = c.toString();
+        Assert.assertTrue(cToString.startsWith("generated.Console"));
+        Assert.assertTrue(cToString.endsWith("[model=PS5, vendor=SONY]"));
+    }
+
+    @Test
+    public void testModel() {
+        Model m = Model.PS_3;
+        Assert.assertEquals(Model.PS_3.value(), m.toString());
+    }
+
+    @Test
+    public void testVendor() {
+        Vendor v = Vendor.MICROSOFT;
+        Assert.assertEquals(Vendor.MICROSOFT.name(), v.toString());
+    }
+}

--- a/jaxb-plugins-parent/tests/enumtostring/src/test/resources/log4j.properties
+++ b/jaxb-plugins-parent/tests/enumtostring/src/test/resources/log4j.properties
@@ -1,0 +1,5 @@
+log4j.rootCategory=DEBUG, stdout
+log4j.appender.stdout=org.apache.log4j.ConsoleAppender
+log4j.appender.stdout.target=system.out
+log4j.appender.stdout.layout=org.apache.log4j.PatternLayout
+log4j.appender.stdout.layout.ConversionPattern=%d %p [%c] - <%m>%n

--- a/jaxb-plugins-parent/tests/pom.xml
+++ b/jaxb-plugins-parent/tests/pom.xml
@@ -19,6 +19,7 @@
     <module>commons_lang</module>
     <module>defaultvalue</module>
     <module>elementwrapper</module>
+    <module>enumtostring</module>
     <module>episodes</module>
     <module>fluentapi</module>
     <module>ignoring</module>

